File: 019.py
'''
How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

1 Jan 1900 was a Monday
'''

'''
This is the brute force way to program it. Is there better way to do it?
'''

def is_leap(year):
    if (year%4== 0 and year%100 != 0) or year%400 == 0 :
         return True
    else:
        return False
        
def output_day(day,month,year):    
#I think this only gives correct output for year > 1900

    days_in_months = [31,28,31,30,31,30,31,31,30,31,30,31]        
    number_of_leap_years = 0  
    total_days = 0
    for i in range(1900,year):
        if is_leap(i) == True:
            number_of_leap_years = number_of_leap_years + 1 
    month_days = sum(days_in_months[:month-1])   
    
    if is_leap(year) == True and month > 2:
        month_days = month_days + 1

    total_days = day + month_days + (year - number_of_leap_years - 1900)*365 + number_of_leap_years*366
    
    day = total_days % 7 
    return day

def print_day(x,y,z):
    day = output_day(x,y,z)
    if day == 0:
        print 'sunday'
    elif day == 1:
        print 'monday'
    elif day == 2 :
        print 'tuesday'
    elif day == 3:
        print 'wednesday'
    elif day == 4:
        print 'thursday'
    elif day == 5:
        print 'friday'
    elif day == 6:
        print 'saturday'

def Problem19():   
    number_of_sundays = 0
    for year in range(1901,2001):
        for month in range(1,13):
            if output_day(1,month,year) == 0:
                number_of_sundays = number_of_sundays + 1      
    print 'Number of Sundays which lie on the first of the months are ' + str(number_of_sundays)

File: 019.hs
{-
1 Jan 1900 = Monday
How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?
-}

leapYears n = sum [isLeap a | a <- [1900..n]]	

isLeap a 
        | (a `mod` 4 == 0) && (a `mod` 100 /= 0) = 1
		| (a `mod` 400 == 0)= 1
		| otherwise = 0


answer = sum [ 1 | y <- [1901..2000] , m <- [1..12] , day 1 m y == 6 ] 

-- d = day, m = month, y = year
day d m y = (d + m_temp + y_temp) `mod` 7
    where
        m_temp = if (isLeap y == 1) then sum $ take (m-1) leap else sum $ take (m-1) noLeap
        y_temp = 366 * (leapYears y') + 365 * (y' -  (leapYears y') -1900)
        y' = y-1

--0 = Monday, 1 = Tuesday ,............, 6 = Sunday

days = ["monday","tuesday","wednesday", "thursday" , "friday","saturday","sunday"]
noLeap = [31,28,31,30,31,30,31,31,30,31,30,31]        
leap =  31 : 29 : drop 2 noLeap

------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------

problem_19 = length . filter (==6) . drop 12 . take 1212 $ since1900
since1900 = scanl nextMonth 0 t
	where 
		t = concat $ replicate 4 noLeap ++ cycle (leap : replicate 3 noLeap)
nextMonth x y = (x + y) `mod` 7

